## Gathering Data ##

First, I'll get the data in the twitter-archive-enhanced.csv file. The file was given by Udacity, so I uploaded it to this notebooks's folder. This is a csv file, so we will read it using pandas.

In [235]:
import pandas as pd
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

Now, I will get the tweet image predictions file programatically. The instructions say that it is available in the URL https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv. I was not sure exactly how to transfer the requested content to a tsv file, so I googled it and used the examples in stack overflow (https://stackoverflow.com/questions/39213597/convert-text-data-from-requests-object-to-dataframe-with-pandas) to help me.

In [236]:
import requests
import io

url_request = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv')
url_content = pd.read_csv(io.StringIO(url_request.text), sep = '\t')
url_content.to_csv('image-predictions.tsv', index = False, sep = '\t')

image_predict = pd.read_csv('image-predictions.tsv', sep = '\t')

The next step is to get the twitter's API data. I followed the steps in the instructions to get my credentials. I will use them here to obtain the data and afther that I will replace them for other strings. To help me import tha data, write it, read it and handle exceptions, I consulted the followint websites: https://towardsdatascience.com/twitter-data-collection-tutorial-using-python-3267d7cfa93e, https://stackoverflow.com/questions/41631873/how-to-read-json-txt-and-parse-json-python-read-txt-python/41632017 and https://stackabuse.com/reading-and-writing-json-to-a-file-in-python.

In [3]:
#Assessing the API, please do not execute this part again because I can't submit the keys to access here
import tweepy

consumer_key = 'I cannot submit this key, please do not execute this part of the code again'
consumer_secret = 'I cannot submit this key, please do not execute this part of the code again'
access_token = 'I cannot submit this key, please do not execute this part of the code again'
access_secret = 'I cannot submit this key, please do not execute this part of the code again'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth)

In [4]:
#Gathering the data from the API, please do not execute this part again because I can't submit the keys to access here
import json
import time

tweet_ids = twitter_archive['tweet_id'].unique()
tweet_data = []
deleted_tweets = []

for id in tweet_ids:
    try:
        tweet_data.append(api.get_status(str(id), tweet_mode='extended')._json)
        print('success')
    except tweepy.RateLimitError:
        time.sleep(15 * 60)
    except tweepy.error.TweepError:
        deleted_tweets.append(str(id))
        print('error')

success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
error
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
success
error
success
success
success
success
success
error
success
success
error
success
success
success
success
success
success
success
success
success
success
success
success
success
error
success
success
success
success
success
success
success
success
su

In [5]:
#Saving the data in json format into a txt file, please do not execute this part again because I can't submit the keys to access here
with open('tweet-json.txt', 'w') as outfile:
    json.dump(tweet_data, outfile)

In [237]:
#Reading the txt file and gathering the important data into a dataframe
with open('tweet-json.txt') as outfile:
    file = json.load(outfile)

    ids = []
    retweet_count = []
    favorite_count = []
for item in file:
    ids.append(item['id'])
    retweet_count.append(item['retweet_count'])
    favorite_count.append(item['favorite_count'])

tweet_counts = pd.DataFrame({'tweet_id': ids, 'retweet_count': retweet_count, 'favorite_count': favorite_count})

## Assessing Data ##

### 1st: Twitter Archive ###

First, let's take a look to the twitter archive echanced file using some functions to understand the data.

In [26]:
twitter_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [27]:
twitter_archive.shape

(2356, 17)

In [28]:
twitter_archive.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1685  681654059175129088                    NaN                  NaN   
1767  678396796259975168                    NaN                  NaN   
195   855138241867124737                    NaN                  NaN   
2252  667801013445750784                    NaN                  NaN   
763   778027034220126208                    NaN                  NaN   

                      timestamp  \
1685  2015-12-29 01:52:46 +0000   
1767  2015-12-20 02:09:34 +0000   
195   2017-04-20 19:16:59 +0000   
2252  2015-11-20 20:25:43 +0000   
763   2016-09-20 00:24:34 +0000   

                                                 source  \
1685  <a href="http://twitter.com/download/iphone" r...   
1767  <a href="http://twitter.com/download/iphone" r...   
195   <a href="http://twitter.com/download/iphone" r...   
2252  <a href="http://twitter.com/download/iphone" r...   
763   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1685  This is Toffee. He's a happy pupper. Appears d...                  NaN   
1767  These little fellas have opposite facial expre...                  NaN   
195   RT @frasercampbell_: oh my... what's that... b...         8.551225e+17   
2252               OMIGOD 12/10 https://t.co/SVMF4Frf1w                  NaN   
763   This is Sophie. She's a Jubilant Bush Pupper. ...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1685                       NaN                        NaN   
1767                       NaN                        NaN   
195               7.475543e+17  2017-04-20 18:14:33 +0000   
2252                       NaN                        NaN   
763                        NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1685  https://twitter.com/dog_rates/status/681654059...                12   
1767  https://twitter.com/dog_rates/status/678396796...                12   
195   https://twitter.com/frasercampbell_/status/855...                14   
2252  https://twitter.com/dog_rates/status/667801013...                12   
763   https://twitter.com/dog_rates/status/778027034...                27   

      rating_denominator    name doggo floofer  pupper puppo  
1685                  10  Toffee  None    None  pupper  None  
1767                  10    None  None    None    None  None  
195                   10    None  None    None    None  None  
2252                  10    None  None    None    None  None  
763                   10  Sophie  None    None  pupper  None

In [29]:
twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [30]:
twitter_archive.tail()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2351  https://twitter.com/dog_rates/status/666049248...                 5   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2353  https://twitter.com/dog_rates/status/666033412...                 9   
2354  https://twitter.com/dog_rates/status/666029285...                 7   
2355  https://twitter.com/dog_rates/status/666020888...                 8   

      rating_denominator  name doggo floofer pupper puppo  
2351                  10  None  None    None   None  None  
2352                  10     a  None    None   None  None  
2353                  10     a  None    None   None  None  
2354                  10     a  None    None   None  None  
2355                  10  None  None    None   None  None

In [31]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [32]:
twitter_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [66]:
image_predict.duplicated().mean()

0.0

#### Problems ####
##### Quality #####
- We can see that the column source has a '< a href =' in the starg of the string and ' rel="nofollow">Twitter for iPhone< /a>' in the end of it, so it does not present the source in a straightforward way. Those parts of the string could be removed.
- We can see, using the info function, that the column expanded_urls has 59 null values (the data has 2.356 rows, and this column has only 2.297 non-null values), which is a quality problem. The columns in_reply_to_status_id and in_reply_to_user_id have a lot of null values. I don't think this is a problem because these columns should be filled only if they are replies. The same happens to retweeted_status_id , retweeted_status_user_id and retweeted_status_timestamp . 
- Timestamp and retweeted_status_timestamp data formats are string and should be datetime. 
- Tweet_id, in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id and retweeted_status_user_id formats are int and should be string.  
- It appears that the maximum values in rating_numerator and rating_denominator were typed wrong. We can see that using the describe function.
- Some tweets could be retweets. We can remove them from all the dataframes using the id's that appear in the retweeted_status_id column.

##### Tydiness ##### 
- Dog "stage" is distributed in four columns in the dataframe (doggo, floofer, pupper, puppo). This could be turned into only one column with the dog's stage information. In case of tweets with more than one dog stage, one of them will nedd to be dropped.

### 2nd: Image Predictions ###

In [39]:
image_predict

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [41]:
image_predict.shape

(2075, 12)

In [42]:
image_predict.sample(5)

tweet_id                                          jpg_url  \
1695  816336735214911488  https://pbs.twimg.com/media/C1Q17WdWEAAjKFO.jpg   
9     666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
1622  803380650405482500  https://pbs.twimg.com/media/CyYub2kWEAEYdaq.jpg   
1080  718234618122661888  https://pbs.twimg.com/media/CfeukpmW4AEGjOE.jpg   
1771  827653905312006145  https://pbs.twimg.com/media/C3xq1ZeWEAEuzw3.jpg   

      img_num                  p1   p1_conf  p1_dog                    p2  \
1695        1  Labrador_retriever  0.919330    True                kuvasz   
9           1    miniature_poodle  0.201493    True              komondor   
1622        1            bookcase  0.890601   False  entertainment_center   
1080        1            malamute  0.370152    True        Siberian_husky   
1771        1              collie  0.285555    True         Border_collie   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
1695  0.049480    True             golden_retriever  0.011934    True  
9     0.192305    True  soft-coated_wheaten_terrier  0.082086    True  
1622  0.019287   False                         file  0.009490   False  
1080  0.356398    True                   Eskimo_dog  0.271042    True  
1771  0.217306    True                Saint_Bernard  0.143245    True

In [43]:
image_predict.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [44]:
image_predict.tail()

tweet_id                                          jpg_url  \
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num           p1   p1_conf  p1_dog                  p2   p2_conf  \
2070        2       basset  0.555712    True    English_springer  0.225770   
2071        1  paper_towel  0.170278   False  Labrador_retriever  0.168086   
2072        1    Chihuahua  0.716012    True            malamute  0.078253   
2073        1    Chihuahua  0.323581    True            Pekinese  0.090647   
2074        1       orange  0.097049   False               bagel  0.085851   

      p2_dog                           p3   p3_conf  p3_dog  
2070    True  German_short-haired_pointer  0.175219    True  
2071    True                      spatula  0.040836   False  
2072    True                       kelpie  0.031379    True  
2073    True                     papillon  0.068957    True  
2074   False                       banana  0.076110   False

In [45]:
image_predict.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [46]:
image_predict.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [59]:
image_predict.duplicated().mean()

0.0

#### Problems ####
##### Quality #####
- tweet_id format is int and should be string, we can see that using info function.
- Some of the tweets are not of dogs, and we should remove them.
- Some of the tweets are retweets and we should remove them.

##### Tydiness ##### 
- This is not exactly a problem, but in order to really use the data as information we are going to have to choose the best model to each tweet based on the confidence interval to determine which is the dog's breed. This will result in two columns: one with the name of the chosen model and one with the dog's breed predicted by that model. After, we can drop the initial model related columns. 

### 3rd: Tweet Counts ###

In [67]:
tweet_counts

id  retweet_count  favorite_count
0     892177421306343426           5435           30092
1     891815181378084864           3584           22646
2     891689557279858688           7467           38003
3     891327558926688256           8020           36285
4     891087950875897856           2689           18296
5     890971913173991426           1733           10660
6     890729181411237888          16327           58499
7     890609185150312448           3733           25207
8     890240255349198849           6302           28726
9     890006608113172480           6341           27742
10    889880896479866881           4309           25236
11    889665388333682689           8631           43246
12    889638837579907072           3859           24325
13    889531135344209921           1946           13697
14    889278841981685760           4602           22719
15    888917238123831296           3875           26295
16    888804989199671297           3664           23077
17    888554962724278272           2976           17799
18    888078434458587136           3000           19674
19    887705289381826560           4674           27339
20    887517139158093824          10166           41823
21    887473957103951883          15527           61904
22    887343217045368832           9114           30372
23    887101392804085760           5153           27687
24    886983233522544640           6588           31309
25    886736880519319552           2737           10780
26    886680336477933568           3866           20290
27    886366144734445568           2725           19067
28    886267009285017600              3             105
29    886258384151887873           5463           25239
...                  ...            ...             ...
2298  666411507551481857            281             385
2299  666407126856765440             30              98
2300  666396247373291520             72             151
2301  666373753744588802             75             165
2302  666362758909284353            488             682
2303  666353288456101888             63             189
2304  666345417576210432            128             258
2305  666337882303524864             80             171
2306  666293911632134144            304             445
2307  666287406224695296             57             128
2308  666273097616637952             69             155
2309  666268910803644416             32              91
2310  666104133288665088           5674           13020
2311  666102155909144576             11              68
2312  666099513787052032             56             139
2313  666094000022159362             65             149
2314  666082916733198337             41             100
2315  666073100786774016            138             278
2316  666071193221509120             52             133
2317  666063827256086533            186             422
2318  666058600524156928             49             103
2319  666057090499244032            116             260
2320  666055525042405380            207             390
2321  666051853826850816            735            1074
2322  666050758794694657             51             119
2323  666049248165822465             37              92
2324  666044226329800704            119             256
2325  666033412701032449             38             105
2326  666029285002620928             41             117
2327  666020888022790149            440            2333

[2328 rows x 3 columns]

In [68]:
tweet_counts.sample(5)

id  retweet_count  favorite_count
1130  724771698126512129            594            2221
397   822859134160621569           2184           12846
1728  678774928607469569            846            2659
164   858843525470990336           3130           14498
1898  674053186244734976           1184            3053

In [69]:
tweet_counts.head()

id  retweet_count  favorite_count
0  892177421306343426           5435           30092
1  891815181378084864           3584           22646
2  891689557279858688           7467           38003
3  891327558926688256           8020           36285
4  891087950875897856           2689           18296

In [70]:
tweet_counts.tail()

id  retweet_count  favorite_count
2323  666049248165822465             37              92
2324  666044226329800704            119             256
2325  666033412701032449             38             105
2326  666029285002620928             41             117
2327  666020888022790149            440            2333

In [71]:
tweet_counts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2328 entries, 0 to 2327
Data columns (total 3 columns):
id                2328 non-null int64
retweet_count     2328 non-null int64
favorite_count    2328 non-null int64
dtypes: int64(3)
memory usage: 54.6 KB


In [72]:
tweet_counts.describe()

id  retweet_count  favorite_count
count  2.328000e+03    2328.000000     2328.000000
mean   7.418640e+17    2559.819158     7245.984966
std    6.819016e+16    4331.591615    11259.215758
min    6.660209e+17       1.000000        0.000000
25%    6.782728e+17     519.000000     1256.500000
50%    7.182408e+17    1195.000000     3138.500000
75%    7.986673e+17    2966.750000     8864.500000
max    8.921774e+17   73372.000000   149150.000000

In [73]:
tweet_counts.duplicated().mean()

0.0

#### Problems ####
##### Quality #####
- id has a int format, it should be string. We can see that using info function.

##### Tydiness ##### 
- This table and the image predictions table could be a part of the tweet_archive dataframe.

## Cleaning data ##

Now, I will do the cleaning process considering the problems cited in the topic above.

### 1st: Twitter Archive ###

First, let's make a copy of the original dataframe:

In [238]:
twitter_archive_clean = twitter_archive.copy()

1. **Quality Issue**: We can see that the column source has a '< a href =' in the starg of the string and ' rel="nofollow">Twitter for iPhone< /a>' in the end of it, so it does not present the source in a straightforward way. Those parts of the string could be removed.

>**Define**: Let's cut the string parts that we do not want to remain. We need to remove 9 characters from the start of the string and 39 characters from the end of the string.

In [239]:
#Code
twitter_archive_clean['source'] = [item[9:-39] for item in twitter_archive_clean['source']]

#Test
twitter_archive_clean['source'].sample(5)

2244    http://twitter.com/download/iphone
2107    http://twitter.com/download/iphone
1572    http://twitter.com/download/iphone
244     http://twitter.com/download/iphone
1307    http://twitter.com/download/iphone
Name: source, dtype: object

Now we can see only the url in the source column, as shown above.

2. **Quality Issue**: We can see, using the info function, that the column expanded_urls has 59 null values (the data has 2.356 rows, and this column has only 2.297 non-null values), which is a quality problem.
>**Define**: In this case we don't have much to do since we don't know what are the missing urls. We could drop the missing value lines, but that would make us lose other important pieces of information of the datarame, inclus=ding more relevant ones than the urls.

3. **Quality Issue**: Timestamp and retweeted_status_timestamp data formats are string and should be datetime. 
4. **Quality Issue**: Tweet_id, in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id and retweeted_status_user_id formats are int and should be string.  
>**Define**: Let's change the data types using astype function.

In [240]:
#Code
twitter_archive_clean['tweet_id'] = twitter_archive_clean['tweet_id'].astype('str')
twitter_archive_clean['in_reply_to_status_id'] = twitter_archive_clean['in_reply_to_status_id'].astype('int', errors = 'ignore').astype('str')
twitter_archive_clean['in_reply_to_user_id'] = twitter_archive_clean['in_reply_to_user_id'].astype('int', errors = 'ignore').astype('str')
twitter_archive_clean['retweeted_status_id'] = twitter_archive_clean['retweeted_status_id'].astype('int', errors = 'ignore').astype('str')
twitter_archive_clean['retweeted_status_user_id'] = twitter_archive_clean['retweeted_status_user_id'].astype('int', errors = 'ignore').astype('str')
twitter_archive_clean['timestamp'] = twitter_archive_clean['timestamp'].astype('datetime64')
twitter_archive_clean['retweeted_status_timestamp'] = twitter_archive_clean['retweeted_status_timestamp'].astype('datetime64')

#Test
print(twitter_archive_clean.info())
twitter_archive_clean.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null object
in_reply_to_status_id         2356 non-null object
in_reply_to_user_id           2356 non-null object
timestamp                     2356 non-null datetime64[ns]
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           2356 non-null object
retweeted_status_user_id      2356 non-null object
retweeted_status_timestamp    181 non-null datetime64[ns]
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: datet

tweet_id in_reply_to_status_id in_reply_to_user_id  \
0  892420643555336193                   nan                 nan   
1  892177421306343426                   nan                 nan   
2  891815181378084864                   nan                 nan   
3  891689557279858688                   nan                 nan   
4  891327558926688256                   nan                 nan   

            timestamp                              source  \
0 2017-08-01 16:23:56  http://twitter.com/download/iphone   
1 2017-08-01 00:17:27  http://twitter.com/download/iphone   
2 2017-07-31 00:18:03  http://twitter.com/download/iphone   
3 2017-07-30 15:58:51  http://twitter.com/download/iphone   
4 2017-07-29 16:00:24  http://twitter.com/download/iphone   

                                                text retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                 nan   
1  This is Tilly. She's just checking pup on you....                 nan   
2  This is Archie. He is a rare Norwegian Pouncin...                 nan   
3  This is Darla. She commenced a snooze mid meal...                 nan   
4  This is Franklin. He would like you to stop ca...                 nan   

  retweeted_status_user_id retweeted_status_timestamp  \
0                      nan                        NaT   
1                      nan                        NaT   
2                      nan                        NaT   
3                      nan                        NaT   
4                      nan                        NaT   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

Now we can see that the columns have the right data types using the table above and the function info.

5. **Quality Issue**: It appears that the maximum values in rating_numerator and rating_denominator were typed wrong.
>**Define**: Let's check for occurencies in which the denominator is not 10 and replace them for 10. In the case of the denominator, let's look for larger numbers than normal and replace them (in the case of th observed maximum numbet, I believe it was supposed to be 17.76.

In [241]:
#Code for denominator
black_list_1 = twitter_archive_clean[twitter_archive_clean['rating_denominator'] != 10]

for i in black_list_1:
    twitter_archive_clean['rating_denominator'].replace(black_list_1[i], 10, inplace = True)

#Test for denominator
twitter_archive_clean['rating_denominator'].describe()    

count    2356.0
mean       10.0
std         0.0
min        10.0
25%        10.0
50%        10.0
75%        10.0
max        10.0
Name: rating_denominator, dtype: float64

In [242]:
#Code for numerator
black_list_2 = twitter_archive_clean[twitter_archive['rating_numerator'] >= 100]['rating_numerator'].unique()

for i in black_list_2:
    if i<1000:
        twitter_archive_clean['rating_numerator'].replace(i, i/10, inplace = True)
    elif i >= 1000:
        twitter_archive_clean['rating_numerator'].replace(i, i/100, inplace = True)

#Test
twitter_archive_clean['rating_numerator'].describe()    

count    2356.000000
mean       11.071842
std         5.163074
min         0.000000
25%        10.000000
50%        11.000000
75%        12.000000
max        99.000000
Name: rating_numerator, dtype: float64

Using the describe functions we can see that in both columns the max value makes more sense now.

6. **Quality Issue**: Some tweets could be retweets. We can remove them from all the dataframes using the id's that appear in the retweeted_status_id column.
>**Define**: I will remove from all dataframes the items whose ids are present in the retweeted_status_id column. I will make a copy of the other two dataframes to do that. After this is done, we won't neet de retweet related columns anymore, so I will drop them.

In [243]:
image_predict_clean = image_predict.copy()
tweet_counts_clean = tweet_counts.copy()

In [245]:
#Code
retweets = [str(int(i)) for i in twitter_archive['retweeted_status_id'].dropna().unique()]

for i in retweets:
    twitter_archive_clean.drop(twitter_archive_clean[twitter_archive_clean['tweet_id'] == i].index, inplace = True)
    image_predict_clean.drop(image_predict_clean[image_predict_clean['tweet_id'].astype('str') == i].index, inplace = True)
    tweet_counts_clean.drop(tweet_counts_clean[tweet_counts_clean['tweet_id'].astype('str') == i].index, inplace = True)

twitter_archive_clean.drop(columns=['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], inplace = True)

#Test
print(twitter_archive_clean.shape[0], twitter_archive.shape[0])
print(image_predict_clean.shape[0], image_predict.shape[0])
print(tweet_counts_clean.shape[0], tweet_counts.shape[0])
twitter_archive_clean.head()

2244 2356
1974 2075
2218 2328


tweet_id in_reply_to_status_id in_reply_to_user_id  \
0  892420643555336193                   nan                 nan   
1  892177421306343426                   nan                 nan   
2  891815181378084864                   nan                 nan   
3  891689557279858688                   nan                 nan   
4  891327558926688256                   nan                 nan   

            timestamp                              source  \
0 2017-08-01 16:23:56  http://twitter.com/download/iphone   
1 2017-08-01 00:17:27  http://twitter.com/download/iphone   
2 2017-07-31 00:18:03  http://twitter.com/download/iphone   
3 2017-07-30 15:58:51  http://twitter.com/download/iphone   
4 2017-07-29 16:00:24  http://twitter.com/download/iphone   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...              13.0   
1  https://twitter.com/dog_rates/status/892177421...              13.0   
2  https://twitter.com/dog_rates/status/891815181...              12.0   
3  https://twitter.com/dog_rates/status/891689557...              13.0   
4  https://twitter.com/dog_rates/status/891327558...              12.0   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

Using the function shape we can see that the twitter_archive_clean had 2,356 records, and now has 2,244. The image_predict_clean dataframe had 2,075 records and now has 1,974. The tweet_counts_clean had 2,328 records and now has 2,218. All this because we removed the retweets. We can also see in the table above that the columns we removed are not there anymore.

7. **Tidyness Issue**: Dog "stage" is distributed in four columns in the dataframe (doggo, floofer, pupper, puppo). This could be turned into only one column with the dog's stage information. In case of tweets with more than one dog stage, one of them will nedd to be dropped.
>**Define**: I am going to use the melt function to create only one column to dog's 'stage', and then I will remove one of the stages for cases with more than one stages reported.

In [246]:
#Code

#First I will create the columns with the dog stage (name of columns of dog stages) and stage (actual stage od that tweet)
twitter_archive_clean = pd.melt(twitter_archive_clean, id_vars=['tweet_id','in_reply_to_status_id','in_reply_to_user_id','timestamp','source','text','expanded_urls','rating_numerator','rating_denominator','name'],
 var_name='dog_stage', value_name='stage')

#Now I will criate an auxiliar array that will receive a random integer number from 1 to 4 if the stage is different of None
# and 5 if the stage is equal to None. I'll add this array to the dataframe and sort the dataframe based on it, using ascending
#order. This way, when I drop the duplicates of ids I will specify to mantain the first occurence, which will be None only if
#there is no other stage reported to that tweet.

import random

aux = []
for i in twitter_archive_clean['stage']:
    if i == 'None':
        aux.append(5)
    else:
        aux.append(random.randint(1,5))

twitter_archive_clean['aux_stage'] = aux

#Sorting values as explained before
twitter_archive_clean.sort_values('aux_stage', ascending=True, inplace = True)

#Dropping duplicates
twitter_archive_clean.drop_duplicates(subset=['tweet_id'], keep='first', inplace = True)

#Removing comlumns that will no longer be used
twitter_archive_clean.drop(columns=['dog_stage', 'aux_stage'], inplace = True)

#Test
twitter_archive_clean.shape[0]  

2244

As the final number of rows of the twitter_archive_clean is 2,244, we can see that we removed all duplicates from the melt process.

### 2nd: Image Predictions ###

1. **Quality Issue**: tweet_id format is int and should be string.
>**Describe**: I will change tweet_id type to string.

In [247]:
#Code
image_predict_clean['tweet_id'] = image_predict_clean['tweet_id'].astype('str')

#Test
image_predict_clean['tweet_id'].describe()

count                   1974
unique                  1974
top       767884188863397888
freq                       1
Name: tweet_id, dtype: object

The describe funtion shows that the tweet_id column is now a string.

2. **Quality Issue**: Some of the tweets are not of dogs, and we should remove them.
3. **Quality Issue**: Some of the tweets are retweets and we should remove them.
4. **Tidyness Issue**: This is not exactly a problem, but in order to really use the data as information we are going to have to choose the best model to each tweet based on the confidence interval to determine which is the dog's breed. This will result in two columns: one with the name of the chosen model and one with the dog's breed predicted by that model. After, we can drop the initial model related columns. 
>**Describe**: First, I will create a column that has the best prediction model in terms of confidence interval, and another column that has the dog's breed chosen by that model. Then, I will remove the rows in which p1_dog or p2_dog or p3_dog (depending on which is the best model) is False, and then I will drop those columns and the model related columns (except for the new ones created) because they won't be used anymore. The retweets were already removed in previous steps.

In [248]:
#Code
#First I will create some auxiliar arrays to get the values of the best models
best_model = []
best_c_interval = []
best_dog_breed = []
best_is_dog = []

#By iterating the dataframe, I get the data of the best model of each row and store it into the auxiliar arrays
for row in image_predict_clean.iterrows():

    aux = max([row[1]['p1_conf'], row[1]['p2_conf'], row[1]['p3_conf']])
    
    if row[1]['p1_conf'] == aux:
        best_model.append('p1')
        best_c_interval.append(row[1]['p1_conf'])
        best_dog_breed.append(row[1]['p1'])
        best_is_dog.append(row[1]['p1_dog'])
    elif row[1]['p2_conf'] == aux:
        best_model.append('p2')
        best_c_interval.append(row[1]['p2_conf'])
        best_dog_breed.append(row[1]['p2'])
        best_is_dog.append(row[1]['p2_dog'])
    else:
        best_model.append('p3')
        best_c_interval.append(row[1]['p3_conf'])
        best_dog_breed.append(row[1]['p3'])
        best_is_dog.append(row[1]['p3_dog'])

#Here, I add the auxiliar arrays to the dataframe
image_predict_clean['best_model'] = best_model
image_predict_clean['dog_breed'] = best_dog_breed
image_predict_clean['is_dog'] = best_is_dog

#Here I drop the rows of tweets that do not refer to dogs
image_predict_clean = image_predict_clean[image_predict_clean['is_dog'] == True]

#Here I drop the columns that we will not use anymore
image_predict_clean.drop(columns=['p1_conf', 'p2_conf', 'p3_conf', 'p1', 'p2', 'p3', 'p1_dog', 'p2_dog', 'p3_dog', 'is_dog'], inplace = True)

#Test
print(image_predict_clean.shape[0])
image_predict_clean.head()
    

1454


tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num best_model               dog_breed  
0        1         p1  Welsh_springer_spaniel  
1        1         p1                 redbone  
2        1         p1         German_shepherd  
3        1         p1     Rhodesian_ridgeback  
4        1         p1      miniature_pinscher

We can see in the table above the final format of our image_predict dataframe after cleaning, exaclty as expected.

## 3rd: Tweet counts ##

1. **Quality Issue**: id has a int format, it should be string.
>**Define**: I will change the id type using astype function

In [249]:
#Code
tweet_counts_clean['tweet_id'] = tweet_counts_clean['tweet_id'].astype('str')

#Test
tweet_counts_clean['tweet_id'].describe()

count                   2218
unique                  2218
top       685315239903100929
freq                       1
Name: tweet_id, dtype: object

The describe funtion shows that the tweet_id column is now a string.

2. **Tidyness Issue**: This table and the image predictions table could be a part of the tweet_archive dataframe.
>**Define**: I will merge the dataframes using the id column. Then, I will drop the rows that have null values for images, because they will not be usefull to our analysis.

In [253]:
#Code
twitter_archive_master = pd.merge(twitter_archive_clean, image_predict_clean,on='tweet_id', how='left')
twitter_archive_master = pd.merge(twitter_archive_master, tweet_counts_clean,on='tweet_id', how='left')
twitter_archive_master.dropna(subset=['jpg_url'], inplace = True)

#Test
print(twitter_archive_master.shape[0])
twitter_archive_master.head()

1454


tweet_id in_reply_to_status_id in_reply_to_user_id  \
0   708845821941387268                   nan                 nan   
5   692752401762250755                   nan                 nan   
7   692905862751522816                   nan                 nan   
11  693262851218264065                   nan                 nan   
12  743222593470234624                   nan                 nan   

             timestamp                              source  \
0  2016-03-13 02:43:08  http://twitter.com/download/iphone   
5  2016-01-28 16:53:37  http://twitter.com/download/iphone   
7  2016-01-29 03:03:25  http://twitter.com/download/iphone   
11 2016-01-30 02:41:58  http://twitter.com/download/iphone   
12 2016-06-15 23:24:09  http://twitter.com/download/iphone   

                                                 text  \
0   Here's a pupper with magic eyes. Not wearing a...   
5   "Hello yes could I get one pupper to go please...   
7   This is Wally. He's being abducted by aliens. ...   
11  I hope you guys enjoy this beautiful snowy pup...   
12  This is a very rare Great Alaskan Bush Pupper....   

                                        expanded_urls  rating_numerator  \
0   https://twitter.com/dog_rates/status/708845821...               9.0   
5   https://twitter.com/dog_rates/status/692752401...              13.0   
7   https://twitter.com/dog_rates/status/692905862...              10.0   
11  https://twitter.com/dog_rates/status/693262851...              11.0   
12  https://twitter.com/dog_rates/status/743222593...              12.0   

    rating_denominator   name   stage  \
0                   10   None  pupper   
5                   10   None  pupper   
7                   10  Wally  pupper   
11                  10   None  pupper   
12                  10      a  pupper   

                                              jpg_url  img_num best_model  \
0     https://pbs.twimg.com/media/CdZTgynWwAATZcx.jpg      1.0         p1   
5   https://pbs.twimg.com/tweet_video_thumb/CZ0mhd...      1.0         p1   
7     https://pbs.twimg.com/media/CZ2yNKhWEAA_7cb.jpg      1.0         p1   
11    https://pbs.twimg.com/media/CZ724fDUYAAytS-.jpg      1.0         p1   
12    https://pbs.twimg.com/media/ClB09z0WYAAA1jz.jpg      1.0         p1   

           dog_breed  retweet_count  favorite_count  
0         schipperke          844.0          2832.0  
5            Samoyed         3443.0          6500.0  
7   Mexican_hairless          825.0          2314.0  
11  golden_retriever          467.0          2099.0  
12            kuvasz         1813.0          5930.0

From the table above we can see that the final dataframe appears to be the way we expected.

## Storing data ##

Now I will store the final dataframe we ceated into a csv file.

In [254]:
twitter_archive_master.to_csv('twitter_archive_master.csv', index=False)